drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2252 @@
+	 *    in push-pull mode.
+	 *  - CLK: Clock is a host to Device signal. CLK operates in push-pull mode.
+	 *  - Data Strobe: Data Strobe is a Device to host signal. Data Strobe operates in
+	 *    push-pull mode."
+	 *
+	 * So, open-drain mode signaling is supported in eMMC as one of the signaling modes for
+	 * the CMD line. But Gen3 and Gen4 boards has MMC/SD controller which is a specialized
+	 * component designed specifically for managing communication with MMC/SD devices. It
+	 * handles low-level operations such as protocol handling, data transfer, and error
+	 * checking and should take care of the low-level details of communicating with the
+	 * MMC/SD card, including setting the bus mode. Moreover, we can use only MMIO mode, the
+	 * processor communicates with the MMC/SD controller through memory read and write
+	 * operations, rather than through dedicated I/O instructions or specialized data transfer
+	 * protocols like SPI or SDIO. Finally, R-Car Gen3 and Gen4 "User’s manuals: Hardware"
+	 * don't have direct configurations for open-drain mode for both PFC and GPIO and Zephyr
+	 * SDHC subsystem doesn't support any bus mode except push-pull.
+	 */
+	if (ios->bus_mode != SDHC_BUSMODE_PUSHPULL) {
+		LOG_ERR("SDHC I/O: not supported bus mode %d", ios->bus_mode);
+		return -ENOTSUP;
+	}
+	host_io->bus_mode = ios->bus_mode;
+
+	/* Set card power */
+	if (ios->power_mode && host_io->power_mode != ios->power_mode) {
+		const struct mmc_rcar_cfg *cfg = dev->config;
+
+		switch (ios->power_mode) {
+		case SDHC_POWER_ON:
+			ret = regulator_enable(cfg->regulator_vmmc);
+			if (ret) {
+				break;
+			}
+
+			k_msleep(data->props.power_delay);
+
+			ret = regulator_enable(cfg->regulator_vqmmc);
+			if (ret) {
+				break;
+			}
+
+			k_msleep(data->props.power_delay);
+			ret = rcar_mmc_enable_clock(dev, true);
+			break;
+		case SDHC_POWER_OFF:
+			if (regulator_is_enabled(cfg->regulator_vqmmc)) {
+				ret = regulator_disable(cfg->regulator_vqmmc);
+				if (ret) {
+					break;
+				}
+			}
+
+			if (regulator_is_enabled(cfg->regulator_vmmc)) {
+				ret = regulator_disable(cfg->regulator_vmmc);
+				if (ret) {
+					break;
+				}
+			}
+
+			ret = rcar_mmc_enable_clock(dev, false);
+			break;
+		default:
+			LOG_ERR("SDHC I/O: not supported power mode %d", ios->power_mode);
+			return -ENOTSUP;
+		}
+
+		if (ret) {
+			return ret;
+		}
+		host_io->power_mode = ios->power_mode;
+	}
+
+	ret = rcar_mmc_set_bus_width(dev, ios);
+	if (ret) {
+		LOG_ERR("SDHC I/O: can't change bus width error %d old %d new %d", ret,
+			host_io->bus_width, ios->bus_width);
+		return ret;
+	}
+
+	ret = rcar_mmc_set_timings(dev, ios);
+	if (ret) {
+		LOG_ERR("SDHC I/O: can't change timing error %d old %d new %d", ret,
+			host_io->timing, ios->timing);
+		return ret;
+	}
+
+	/* Set host driver type */
+	if (ios->driver_type && ios->driver_type != host_io->driver_type) {
+		switch (ios->driver_type) {
```

Comment:
Why have this switch statement? It seems like the `sdhc_set_io` call will always return an error if the user tries to change the driver type

Reply:
Maybe I'm missing something here, sorry- if the user calls `sdhc_set_io` with a new driver type, wouldn't this switch statement fall through to the error? Or are you saying the SDMMC stack would never call `sdhc_set_io` with a changed driver type, since it will check the support fields?